vm/state_migration/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+        migrations.insert(
+            *actorv3::POWER_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::POWER_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::SYSTEM_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::SYSTEM_ACTOR_CODE_ID),
+        );
+        migrations.insert(
+            *actorv3::VERIFREG_ACTOR_CODE_ID,
+            nil_migrator_v4(*actorv4::VERIFREG_ACTOR_CODE_ID),
+        );
+
+        Self {
+            migrations,
+            deferred_code_ids: HashSet::new(),
+        }
+    }
+}
+
+impl<BS: BlockStore + Send + Sync> StateMigration<BS> {
+    pub fn add_migrator(&mut self, prior_cid: Cid, migrator: Migrator<BS>) {
+        self.migrations.insert(prior_cid, migrator);
+    }
+
+    pub fn migrate_state_tree(
+        &mut self,
+        store: Arc<BS>,
+        prior_epoch: ChainEpoch,
+        actors_in: StateTree<BS>,
+        mut actors_out: StateTree<BS>,
+    ) -> MigrationResult<Cid> {
+        if self.migrations.len() + self.deferred_code_ids.len() != ACTORS_COUNT {
+            return Err(MigrationError::IncompleteMigrationSpec(
+                self.migrations.len(),
+            ));
+        }
+
+        let cpus = num_cpus::get();
+        let chan_size = 2;
+
+        log::info!(
+            "Using {} CPUs for migration and channel size of {}",
+            cpus,
+            chan_size
+        );
+
+        let pool = rayon::ThreadPoolBuilder::new()
+            .thread_name(|id| format!("nv12 migration thread: {}", id))
+            .num_threads(cpus)
+            .build()
+            .map_err(MigrationError::ThreadPoolCreation)?;
+
+        let (state_tx, state_rx) = crossbeam_channel::bounded(chan_size);
+        let (job_tx, job_rx) = crossbeam_channel::bounded(chan_size);
+
+        pool.scope(|s| {
+            let store_clone = store.clone();
+
+            s.spawn(move |_| {
+                actors_in
+                    .for_each(|addr, state| {
+                        state_tx
+                            .send((addr, state.clone()))
+                            .expect("failed sending actor state through channel");
+                        Ok(())
+                    })
+                    .expect("Failed iterating over actor state");
+            });
+
+            s.spawn(move |scope| {
+                while let Ok((address, state)) = state_rx.recv() {
+                    let job_tx = job_tx.clone();
+                    let store_clone = store_clone.clone();
+                    let migrator = self.migrations.get(&state.code).cloned().unwrap();
+                    scope.spawn(move |_| {
+                        let job = MigrationJob {
+                            address,
+                            actor_state: state,
+                            actor_migration: migrator,
+                        };
+
+                        let job_output = job.run(store_clone, prior_epoch).unwrap_or_else(|e| {
+                            panic!(
```

Comment:
is it okay to panic here?... i dont expect there be a good way to handle a state migration failure, but i wonder if we can do this more cleanly

Reply:
Yeah, had thought about  it, I could have created a global error variable outside and assign any `Err`s and could bail early from the spawned thread. But i guess it is safe to panic here as it's a child thread and not proceed with any further migration state change. Totally open if we have better error handling ideas.

Reply:
I think we need to ask whether is it safe for the migration to continue if any of the spawned thread fails, or we should discard the whole migration. If it's the later case, then yeah we can bail early with an `Err`. Thoughts?